### PR TITLE
Enforce at least one device group when creating a campaign.

### PR DESCRIPTION
**Pull request by houcros**

### CODE REVIEW

File: src/main/scala/com/advancedtelematic/campaigner/http/CampaignResource.scala [hunk 1]
```diff
@@ -58,7 +58,8 @@ class CampaignResource(extractAuth: Directive1[AuthedNamespaceScope],
             complete(campaigns.allCampaigns(ns, offset, limit))
           } ~
           (post & entity(as[CreateCampaign])) { request =>
-            complete(StatusCodes.Created -> createCampaign(ns, request))
+            if (request.groups.isEmpty) complete(BadRequest -> "A campaign must include at least one device group.")
```
[the remark this answers is not in the document]
Comment (houcros) [reply src/main/scala/com/advancedtelematic/campaigner/http/CampaignResource.scala]: Done :)

File: src/main/scala/com/advancedtelematic/campaigner/http/CampaignResource.scala [hunk 2]
```diff
@@ -58,7 +59,8 @@ class CampaignResource(extractAuth: Directive1[AuthedNamespaceScope],
             complete(campaigns.allCampaigns(ns, offset, limit))
           } ~
           (post & entity(as[CreateCampaign])) { request =>
-            complete(StatusCodes.Created -> createCampaign(ns, request))
```
Comment (simao) [inline src/main/scala/com/advancedtelematic/campaigner/http/CampaignResource.scala]: Can you try using NonEmptyList CreateCampaign ? I think you might get this for free from circe.

Needs rebase

File: src/test/scala/com/advancedtelematic/campaigner/data/Generators.scala [hunk 3]
```diff
@@ -38,7 +38,7 @@ object Generators {
   val genCreateCampaign: Gen[CreateCampaign] = for {
     n   <- arbitrary[String]
     update <- arbitrary[UpdateId]
-    gs  <- arbitrary[Set[GroupId]]
+    gs  <- arbitrary[Set[GroupId]].retryUntil(_.nonEmpty)
```
Comment (simao) [inline src/test/scala/com/advancedtelematic/campaigner/data/Generators.scala]: You should change the generator in Generators instead, look for Gen.listOf methods.

File: src/main/scala/com/advancedtelematic/campaigner/http/CampaignResource.scala (hunk 2 above)
Comment (houcros) [reply src/main/scala/com/advancedtelematic/campaigner/http/CampaignResource.scala]: Alright, so in `CreateCampaign` I tried turning the `Set[GrouId]` to a `NonEmptySet[GroupId]`, but the decoders don't work anymore and I can't find a way without implementing the whole encoder/decoder.
Any tips?

Comment (koshelev) [reply src/main/scala/com/advancedtelematic/campaigner/http/CampaignResource.scala]: You just need a decoder for `NonEmptySet[T]` given there is a `Decoder[T]`. Me or Simao can help you with it if you don't know how. It is quite simple

Comment (simao) [reply src/main/scala/com/advancedtelematic/campaigner/http/CampaignResource.scala]: I was almost sure circe came with that decoder though

Comment (koshelev) [reply src/main/scala/com/advancedtelematic/campaigner/http/CampaignResource.scala]:
```
def nesDecoder[T](implicit dec: Decoder[Set[T]]): Decoder[NonEmptySet[T]] =
    dec.emapTry(xs => if(xs.isEmpty) Failure(DecodingFailure("No elements")) else Success(xs))
```

Comment (simao) [reply src/main/scala/com/advancedtelematic/campaigner/http/CampaignResource.scala]: Needs NonEmptySet.fromSetUnsafe() or unpack option from NonEmptySet.fromSet() ;)

Comment (koshelev) [reply src/main/scala/com/advancedtelematic/campaigner/http/CampaignResource.scala]: Actually, it should be 
```
    def nesDecoder[T](implicit dec: Decoder[Set[T]], ord: Ordering[T]): Decoder[NonEmptySet[T]] = {
    import scala.collection.immutable.SortedSet
    dec.emapTry(xs => if(xs.isEmpty) Failure(DecodingFailure("No elements", Nil)) else Success(NonEmptySet.fromSetUnsafe(xs.to[SortedSet])))
  }
```

Comment (simao) [reply src/main/scala/com/advancedtelematic/campaigner/http/CampaignResource.scala]: let's open a circe pr :+1:

Comment (koshelev) [reply src/main/scala/com/advancedtelematic/campaigner/http/CampaignResource.scala]: We could, but will need to do bit more work to get rid of `Nil` in `DecodingFailure("No elements", Nil)`. And to go through internal OSS contribution process :)

Comment (houcros) [reply src/main/scala/com/advancedtelematic/campaigner/http/CampaignResource.scala]: I'll create this codecs in libats.

Comment (houcros) [reply src/main/scala/com/advancedtelematic/campaigner/http/CampaignResource.scala]: Alright, in the end I refactored this to use `NonEmptyList`. WDYT?

File: src/test/scala/com/advancedtelematic/campaigner/data/Generators.scala (hunk 3 above)
Comment (houcros) [reply src/test/scala/com/advancedtelematic/campaigner/data/Generators.scala]: I changed this generator now.

File: src/main/scala/com/advancedtelematic/campaigner/http/CampaignResource.scala (hunk 2 above)
Comment (koshelev) [reply src/main/scala/com/advancedtelematic/campaigner/http/CampaignResource.scala]: I think `Set` still makes sense. It does not make sense to have a group several times in this collection

Comment (houcros) [reply src/main/scala/com/advancedtelematic/campaigner/http/CampaignResource.scala]: I didn't change `CampaignRepository#persist`, that still takes a `Set[GroupId]`.

File: src/main/scala/com/advancedtelematic/campaigner/data/DataType.scala [hunk 4]
```diff
@@ -170,4 +173,7 @@ object DataType {
     val DESCRIPTION, ESTIMATED_INSTALLATION_DURATION, ESTIMATED_PREPARATION_DURATION = Value
   }
 
+  implicit def nonEmptyListToSet[A](nel: NonEmptyList[A]): Set[A] = {
```
Comment (koshelev) [inline src/main/scala/com/advancedtelematic/campaigner/data/DataType.scala]: Please do it explicitly

Comment (houcros) [reply src/main/scala/com/advancedtelematic/campaigner/data/DataType.scala]: Doing explicitly now :)

**Comment by jerrytrieu** [comment]
Avoid conflict with #62

**Comment by simao** [comment]
what happened to this?

**Comment by houcros** [comment]
Moved to #77.
I renamed the branch, the hat-box deployments were failing because there was no city in the branch name.

**Pull request closed without merging**